nix/nixos/cardano-wallet-service.nix

```diff
@@ -0,0 +1,193 @@
+        Default to the set of environments found at
+        iohk-nix/cardano-lib/default.nix, or, it must be an attribute set
+        of the same form.
+      '';
+    };
+
+    walletMode = mkOption {
+      type = types.enum [ "mainnet" "staging" "testnet" ];
+      default =
+        if cfg.environment == "mainnet"
+        then "mainnet"
+        else if cfg.environment == "staging"
+        then "staging"
+        else "testnet";
+      description = "Which mode to start wallet in: --mainnet, --staging or --testnet";
+    };
+
+    database = mkOption {
+      type = types.nullOr types.str;
+      default = "$STATE_DIRECTORY";
+      description = ''Directory for storing wallets. Run in-memory if null.'';
+    };
+
+    syncTolerance = mkOption {
+      type = types.ints.positive;
+      default = 300;
+      description = "Time duration within which we consider being synced with the network. Expressed in seconds.";
+    };
+
+    poolMetadataFetching = mkOption {
+      type = types.enum [ "none" "direct" "smash" ];
+      default = if cardanoEnvAttrs ? smashUrl then "smash" else "none";
+      description = ''
+        Sets the stake pool metadata fetching strategy.
+        "smash" will use the 'smashUrl' attribute of the 'environnement'
+        as metadata proxy server (default if exists).
+        Use "direct" to fetch directly from the
+        registered pool URLs, or "none" to completely disable
+        stake pool metadata (default if no 'smashUrl' defined).'';
+    };
+
+    tokenMetadataServer = mkOption {
+      type = types.nullOr types.str;
+      default = cardanoEnvAttrs.metadataUrl or null;
+      description = ''
+        Sets the URL of the token metadata server,
+        default to 'metadataUrl' of the 'environnement' attribute, if exists.
+        If unset, metadata will not be fetched. By using this
+        option, you are fully trusting the operator of the
+        metadata server to provide authentic token metadata.
+      '';
+    };
+
+    logLevel = mkOption {
+      type = logLevels;
+      default = "info";
+      description = "Global minimum severity for a message to be logged.";
+    };
+
+    trace = mkOption {
+      type = types.attrsOf logLevels;
+      default = { };
+      description = ''
+        For each tracer, minimum severity for a message to be logged, or
+        "off" to disable the tracer".
+      '';
+    };
+
+    rtsArgs = mkOption {
+      type = types.listOf types.str;
+      default = [ "-N2" ];
+      apply = args:
+        if args != [ ] then
+          [ "+RTS" ] ++ args ++ [ "-RTS" ]
+        else [ ];
+      description = ''Extra CLI args for 'cardano-wallet', to be surrounded by "+RTS"/"-RTS"'';
+    };
+
+  };
+
+  config = mkIf cfg.enable {
+
+    systemd.services.cardano-wallet = {
+      description = "cardano-wallet daemon";
+      wantedBy = [ "multi-user.target" ];
+      inherit (cfg) script;
+
+      serviceConfig = {
+        DynamicUser = true;
```

Comment:
Will this user have permission to open the cardano-node socket file?
We really want it so that the socket file is protected by permissions, and only members of the `cardano-node` group (for example) can open it.
But how would someone add the cardano-wallet service's user to the correct group?

Reply:
`serviceConfig.SupplementaryGroups = "cardano-node";` should work.